autoload/lsp/ui/vim/semantic.vim

```diff
@@ -0,0 +1,187 @@
+        let l:scopes = lsp#ui#vim#semantic#get_scopes(a:server)
+        for l:scope_idx in range(len(l:scopes))
+            let l:scope = l:scopes[l:scope_idx]
+            let l:hl = s:get_hl_name(a:server, l:scope)
+
+            silent! call prop_type_add(s:get_textprop_name(a:server, l:scope_idx), {'bufnr': a:buf, 'highlight': l:hl, 'combine': v:true})
+        endfor
+    endif
+
+    call setbufvar(a:buf, 'lsp_did_semantic_setup', 1)
+endfunction
+
+function! s:add_highlight(server, buf, line, highlights) abort
+    " Skip if the buffer doesn't exist. This might happen when a buffer is
+    " opened and quickly deleted.
+    if !bufexists(a:buf) | return | endif
+
+    let l:scopes = lsp#ui#vim#semantic#get_scopes(a:server)
+
+    if s:use_vim_textprops
+        " Clear text properties from the previous run
+        for l:scope_idx in range(len(l:scopes))
+            let l:scope = l:scopes[l:scope_idx]
+            call prop_remove({'bufnr': a:buf, 'type': s:get_textprop_name(a:server, l:scope_idx), 'all': v:true}, a:line + 1)
+        endfor
+
+        for l:highlight in a:highlights
+            call prop_add(a:line + 1, l:highlight['char'] + 1, { 'length': l:highlight['length'], 'bufnr': a:buf, 'type': s:get_textprop_name(a:server, l:highlight['scope'])})
+        endfor
+    elseif s:use_nvim_highlight
+        " Clear text properties from the previous run
+        call nvim_buf_clear_namespace(a:buf, s:namespace_id, a:line, a:line + 1)
+
+        for l:highlight in a:highlights
+            call nvim_buf_add_highlight(a:buf, s:namespace_id, s:get_hl_name(a:server, l:scopes[l:highlight['scope']]), a:line, l:highlight['char'], l:highlight['char'] + l:highlight['length'])
+        endfor
+    endif
+endfunction
+
+function! s:get_hl_name(server, scope) abort
+    let l:hl = 'Ignore'
+
+    try
+        let l:info = lsp#get_server_info(a:server)
+        let l:hl = l:info['semantic_highlight']
+        let l:i = 0
+
+        while (l:i < len(a:scope)) && has_key(l:hl, a:scope[i])
+            let l:hl = l:hl[a:scope[i]]
+            let l:i += 1
+        endwhile
+    catch
+    endtry
+
+    return type(l:hl) == type('') ? l:hl : 'Ignore'
+endfunction
+
+function! s:get_textprop_name(server, scope_index) abort
+    return 'vim-lsp-semantic-' . a:server . '-' . a:scope_index
+endfunction
+
+" Response parsing functions {{{1
+
+" Converts a list of bytes (MSB first) to a Number.
+function! s:octets_to_number(octets) abort
+    let l:ret = 0
+
+    for l:octet in a:octets
+        let l:ret *= 256
+        let l:ret += l:octet
+    endfor
+
+    return l:ret
+endfunction
+
+function! s:tokens_to_hl_info(token) abort
+    let l:ret = []
+    let l:octets = lsp#utils#base64_decode(a:token)
+
+    for l:i in range(0, len(l:octets) - 1, 8)
+        let l:char = s:octets_to_number(l:octets[i : i+3])
+        let l:length = s:octets_to_number(l:octets[i+4 : i+5])
+        let l:scope = s:octets_to_number(l:octets[i+6 : i+7])
```

Comment:
This is a point to the code. Just question. Where this octets come from?

Reply:
Do you mean why I used these specific offsets?

Reply:
Yes, it is.

Reply:
I don't understand why octet format is required in this part.

Reply:
This is what the spec says, see "tokens" in https://github.com/microsoft/vscode-languageserver-node/blob/5a9b33c23de84c3341e011e79221795a8059375b/protocol/src/protocol.semanticHighlighting.proposed.md.
The tokens contain the character (4 bytes), length (2 bytes), and the scope (2 bytes). This is done for all tokens, and the resulting number is base64 encoded. Not sure why they made this process so complicated, but we'll have to work with it, I guess.

Reply:
Ah, now I understand why I couldn't find this from the Language Server Specification. Okay, please wait. I'll try this with some language servers. Thanks.

Reply:
So far, I've only found eclipse LS. `ccls` seems to use their own implementation.